src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs

```diff
@@ -0,0 +1,324 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+#if NET
+using System.Runtime.CompilerServices;
+#endif
+
+namespace Microsoft.Extensions.Diagnostics.Metrics
+{
+    internal class ListenerSubscription : IObservableInstrumentsSource, IDisposable
+    {
+        private readonly MeterListener _meterListener;
+        private readonly IMetricsListener _metricsListener;
+#if NET
+        private readonly ConditionalWeakTable<Instrument, InstrumentStatus> _instruments = new();
+#else
+        // TODO: Can't enumerate ConditionalWeakTable in .NET Standard 2.0. How do we clean up the instruments?
+        private readonly Dictionary<Instrument, InstrumentStatus> _instruments = new();
+#endif
+        private IList<InstrumentRule> _rules = Array.Empty<InstrumentRule>();
+        private bool _disposed;
+
+        internal ListenerSubscription(IMetricsListener metricsListener)
+        {
+            _metricsListener = metricsListener;
+
+            _meterListener = new MeterListener();
+        }
+
+        public void Initialize()
+        {
+            _meterListener.InstrumentPublished = InstrumentPublished;
+            _meterListener.MeasurementsCompleted = MeasurementsCompleted;
+            var handlers = _metricsListener.GetMeasurementHandlers();
+            _meterListener.SetMeasurementEventCallback(handlers.ByteHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.ShortHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.IntHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.LongHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.FloatHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.DoubleHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.DecimalHandler);
+            _meterListener.Start();
+            _metricsListener.Initialize(this);
+        }
+
+        private void InstrumentPublished(Instrument instrument, MeterListener _)
+        {
+            lock (_instruments)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                if (_instruments.TryGetValue(instrument, out var _))
```

Comment:
>(_instruments.TryGetValue(instrument, out var _)) [](http://example.com/codeflow?start=19&length=49)

Should we ensure the dictionary is created with reference equality comparer? In the future there may be added equality support to the instrument which possibly can make two instruments with the same name/descriptions be equal.

Reply:
That's not available on all of our TFMs.

Reply:
Can polyfill this I guess the same way we do like https://github.com/dotnet/runtime/blob/79b970739df6ff8d75ff7f4f262a8b2785c20940/src/libraries/Microsoft.Extensions.Options.DataAnnotations/src/Microsoft.Extensions.Options.DataAnnotations.csproj#L15